File: ios/Plugin/MicrophonePlugin.swift
import Capacitor
import AVFoundation

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(MicrophonePlugin)
public class MicrophonePlugin: CAPPlugin {
    private var implementation: Microphone?

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        var result: [String: Any] = [:]
        for permission in MicrophonePermissionType.allCases {
            let state: String
            switch permission {
            case .microphone:
                state = AVCaptureDevice.authorizationStatus(for: .audio).authorizationState
            }
            result[permission.rawValue] = state
        }
        call.resolve(result)
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        // TODO: (CHECK) We are not even sending permission list (Do we need it ?)
        // get the list of desired types, if passed
        let typeList = call.getArray("permissions", String.self)?.compactMap({ (type) -> MicrophonePermissionType? in
            return MicrophonePermissionType(rawValue: type)
        }) ?? []
        // otherwise check everything
        let permissions: [MicrophonePermissionType] = (typeList.count > 0) ? typeList : MicrophonePermissionType.allCases
        // request the permissions
        let group = DispatchGroup()
        for permission in permissions {
            switch permission {
            case .microphone:
                group.enter()
                AVCaptureDevice.requestAccess(for: .audio) { _ in
                    group.leave()
                }
            }
        }
        group.notify(queue: DispatchQueue.main) { [weak self] in
            self?.checkPermissions(call)
        }
    }

    @objc func startRecording(_ call: CAPPluginCall) {
        if !isAudioRecordingPermissionGranted() {
            call.reject(StatusMessageTypes.microphonePermissionNotGranted.rawValue)
            return
        }

        if implementation != nil {
            call.reject(StatusMessageTypes.recordingInProgress.rawValue)
            return
        }

        implementation = Microphone()
        if implementation == nil {
            call.reject(StatusMessageTypes.cannotRecordOnThisPhone.rawValue)
            return
        }

        let successfullyStartedRecording = implementation!.startRecording()
        if successfullyStartedRecording == false {
            call.reject(StatusMessageTypes.cannotRecordOnThisPhone.rawValue)
        } else {
            call.resolve(["status": StatusMessageTypes.recordingStared.rawValue])
        }
    }

    @objc func pauseRecording(_ call: CAPPluginCall) {
        if implementation == nil {
            call.reject(StatusMessageTypes.noRecordingInProgress.rawValue)
            return
        }

        implementation?.pauseRecording()
        call.resolve(["status": StatusMessageTypes.recordingPaused.rawValue])
    }

    @objc func resumeRecording(_ call: CAPPluginCall) {
        if implementation == nil {
            call.reject(StatusMessageTypes.noRecordingInProgress.rawValue)
            return
        }

        implementation?.resumeRecording()
        call.resolve(["status": StatusMessageTypes.recordingResumed.rawValue])
    }

    @objc func getCurrentStatus(_ call: CAPPluginCall) {
        if let impl = implementation {
            call.resolve(["status": impl.getCurrentStatus()])
        } else {
            call.resolve(["status": StatusMessageTypes.noRecordingInProgress.rawValue])
        }
    }

    @objc func stopRecording(_ call: CAPPluginCall) {
        if implementation == nil {
            call.reject(StatusMessageTypes.noRecordingInProgress.rawValue)
            return
        }

        implementation?.stopRecording()

        let audioFileUrl = implementation?.getOutputFile()
        if audioFileUrl == nil {
            implementation = nil
            call.reject(StatusMessageTypes.failedToFetchRecording.rawValue)
            return
        }

        let webURL = bridge?.portablePath(fromLocalURL: audioFileUrl)

        let audioRecording = AudioRecording(
            path: audioFileUrl?.absoluteString,
            webPath: webURL?.path,
            duration: getAudioFileDuration(audioFileUrl),
            format: ".m4a",
            mimeType: "audio/aac"
        )
        implementation = nil
        if audioRecording.duration < 0 {
            call.reject(StatusMessageTypes.failedToFetchRecording.rawValue)
        } else {
            call.resolve(audioRecording.toDictionary())
        }
    }

    private func isAudioRecordingPermissionGranted() -> Bool {
        return AVAudioSession.sharedInstance().recordPermission == AVAudioSession.RecordPermission.granted
    }

    private func getAudioFileDuration(_ filePath: URL?) -> Int {
        if filePath == nil {
            return -1
        }
        return Int(CMTimeGetSeconds(AVURLAsset(url: filePath!).duration) * 1000)
    }
}

File: android/src/main/java/com/mozartec/capacitor/microphone/MicrophonePlugin.java
package com.mozartec.capacitor.microphone;

import android.Manifest;
import android.media.MediaPlayer;
import android.net.Uri;
import android.util.Log;
import com.getcapacitor.FileUtils;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import java.io.File;
import java.util.List;
import org.json.JSONException;

@CapacitorPlugin(
    name = "Microphone",
    permissions = { @Permission(strings = { Manifest.permission.RECORD_AUDIO }, alias = MicrophonePlugin.MICROPHONE) }
)
public class MicrophonePlugin extends Plugin {

    // Permission alias constants
    static final String MICROPHONE = "microphone";

    private Microphone implementation;

    // Looks like checkPermissions is available out of the box

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        // Save the call to be able to access it in microphonePermissionsCallback
        bridge.saveCall(call);
        // If the microphone permission is defined in the manifest, then we have to prompt the user
        // or else we will get a security exception when trying to present the microphone. If, however,
        // it is not defined in the manifest then we don't need to prompt and it will just work.
        if (isPermissionDeclared(MICROPHONE)) {
            // just request normally
            super.requestPermissions(call);
        } else {
            // the manifest does not define microphone permissions, so we need to decide what to do
            // first, extract the permissions being requested
            // TODO: (CHECK) We are not even sending permission list (Do we need it ?)
            JSArray providedPerms = call.getArray("permissions");
            List<String> permsList = null;
            try {
                permsList = providedPerms.toList();
            } catch (JSONException e) {}

            // TODO: (CHECK) This may not even be needed as till now we only need mic permission
            if (permsList != null && permsList.size() == 1 && permsList.contains(MICROPHONE)) {
                // the only thing being asked for was the microphone so we can just return the current state
                checkPermissions(call);
            } else {
                // we need to ask about microphone so request storage permissions
                // This will break complaining about permission missing in manifest
                requestPermissionForAlias(MICROPHONE, call, "checkPermissions");
            }
        }
    }

    @PermissionCallback
    private void microphonePermissionsCallback(PluginCall call) {
        checkPermissions(call);
    }

    @PluginMethod
    public void startRecording(PluginCall call) {
        if (!isAudioRecordingPermissionGranted()) {
            call.reject(StatusMessageTypes.MicrophonePermissionNotGranted.getValue());
            return;
        }

        if (implementation != null) {
            call.reject(StatusMessageTypes.RecordingInProgress.getValue());
            return;
        }

        try {
            implementation = new Microphone(getContext());
            implementation.startRecording();
            String status = StatusMessageTypes.RecordingStared.getValue();
            emitStatus(status);
            JSObject success = new JSObject();
            success.put("status", status);
            call.resolve(success);
        } catch (Exception exp) {
            call.reject(StatusMessageTypes.CannotRecordOnThisPhone.getValue());
        }
    }

    @PluginMethod
    public void pauseRecording(PluginCall call) {
        if (implementation == null) {
            call.reject(StatusMessageTypes.NoRecordingInProgress.getValue());
            return;
        }

        try {
            implementation.pauseRecording();
            String status = StatusMessageTypes.RecordingPaused.getValue();
            emitStatus(status);
            JSObject success = new JSObject();
            success.put("status", status);
            call.resolve(success);
        } catch (Exception exp) {
            call.reject(StatusMessageTypes.RecordingFailed.getValue());
        }
    }

    @PluginMethod
    public void resumeRecording(PluginCall call) {
        if (implementation == null) {
            call.reject(StatusMessageTypes.NoRecordingInProgress.getValue());
            return;
        }

        try {
            implementation.resumeRecording();
            String status = StatusMessageTypes.RecordingResumed.getValue();
            emitStatus(status);
            JSObject success = new JSObject();
            success.put("status", status);
            call.resolve(success);
        } catch (Exception exp) {
            call.reject(StatusMessageTypes.RecordingFailed.getValue());
        }
    }

    @PluginMethod
    public void getCurrentStatus(PluginCall call) {
        JSObject result = new JSObject();
        String status;
        if (implementation == null) {
            status = StatusMessageTypes.NoRecordingInProgress.getValue();
        } else {
            status = implementation.getCurrentStatus();
        }
        result.put("status", status);
        emitStatus(status);
        call.resolve(result);
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void removeStatusListener(PluginCall call) {
        removeListener(call);
        call.resolve();
    }

    @PluginMethod
    public void stopRecording(PluginCall call) {
        if (implementation == null) {
            call.reject(StatusMessageTypes.NoRecordingInProgress.getValue());
            return;
        }

        try {
            implementation.stopRecording();
            File audioFileUrl = implementation.getOutputFile();
            Uri newUri = Uri.fromFile(audioFileUrl);
            String webURL = FileUtils.getPortablePath(getContext(), bridge.getLocalUrl(), newUri);
            Log.e("webURL", webURL);
            int duration = getAudioFileDuration(audioFileUrl.getAbsolutePath());
            Log.e("duration", duration + "");
            Log.e("newUri", newUri.toString());
            Recording recording = new Recording(newUri.toString(), webURL, duration, ".m4a", "audio/aac");
            if (duration < 0) {
                call.reject(StatusMessageTypes.FailedToFetchRecording.getValue());
            } else {
                emitStatus(StatusMessageTypes.NoRecordingInProgress.getValue());
                call.resolve(recording.toJSObject());
            }
        } catch (Exception exp) {
            call.reject(StatusMessageTypes.FailedToFetchRecording.getValue());
        } finally {
            implementation = null;
        }
    }

    private boolean isAudioRecordingPermissionGranted() {
        return getPermissionState(MICROPHONE) == PermissionState.GRANTED;
    }

    private int getAudioFileDuration(String filePath) {
        try {
            MediaPlayer mp = new MediaPlayer();
            mp.setDataSource(filePath);
            mp.prepare();
            return mp.getDuration();
        } catch (Exception ignore) {
            return -1;
        }
    }

    private void emitStatus(String status) {
        JSObject ret = new JSObject();
        ret.put("status", status);
        notifyListeners("status", ret);
    }
}
